### PR TITLE
DEP: Deprecate inexact matches for mode, searchside

### PR DESCRIPTION
**Pull request by anirudh2290**
Depends on https://github.com/numpy/numpy/pull/16009 

Please see the discussion here : https://github.com/numpy/numpy/pull/16007#discussion_r410517462

### CODE REVIEW

File: numpy/core/tests/test_conversion_utils.py [hunk 1]
```diff
@@ -107,18 +109,23 @@ class TestSearchsideConverter(StringConverterTestCase):
...
 class TestOrderConverter(StringConverterTestCase):
     """ Tests of PyArray_OrderConverter """
     conv = mt.run_order_converter
     def test_valid(self):
-        self._check('c', 'NPY_CORDER')
-        self._check('f', 'NPY_FORTRANORDER')
-        self._check('a', 'NPY_ANYORDER')
-        self._check('k', 'NPY_KEEPORDER')
+        with suppress_warnings() as sup:
+            sup.filter(DeprecationWarning, "case insensitive matches")
```
Comment (eric-wieser) [inline numpy/core/tests/test_conversion_utils.py:124]: This should be inside `self._check`

**Comment by eric-wieser** [comment]
Think you had a typo in the PR title, corrected it.

File: numpy/core/tests/test_conversion_utils.py (hunk 1 above)
Comment (anirudh2290) [reply numpy/core/tests/test_conversion_utils.py:124]: Thanks @eric-wieser  for the review! i was considering that, but i didnt want this to suppress unintended warnings when TestSortKindConverter or TestSelectkindConverter is used. Since we are already filtering on DeprecatedWarning with specific message its probably better to put it back in self._check.

Comment (eric-wieser) [reply numpy/core/tests/test_conversion_utils.py:124]: If you put it in `self.check`, you can make it conditional on the value of `self.exact_match`.

File: numpy/core/tests/test_multiarray.py [hunk 2]
```diff
@@ -35,6 +35,10 @@
 # Need to test an object that does not fully implement math interface
 from datetime import timedelta, datetime
 
+suppress_inexact_match = suppress_warnings()
+suppress_inexact_match.filter(DeprecationWarning,
+                              "inexact matches and case insensitive matches")
```
Comment (seberg) [inline numpy/core/tests/test_multiarray.py:40]: Its OK to use this, but I think nowadays we can just use the pytest warning ignore hooks. I assume those works well. The `suppress_warnings()` solved a real issue, but I think python fixed that mostly by now. Now it would still add better scoping, but we never really need that, so its probably just unnecessary to use this...

Comment (anirudh2290) [reply numpy/core/tests/test_multiarray.py:40]: Thanks for bringing this up @seberg ! Before using suppress_warnings I initially tried `pytest.mark.filterwarnings("ignore::DeprecationWarning")` for example for the _check function in test_conversion_utils but it didnt suppress the error for some reason. Also I saw this decorator sparsely used with the codebase and the places it had used were xfail tests. I didnt dig more into it but switched to suppress_warning. I can look more into this if we are trying to stick to it in the future.

Comment (seberg) [reply numpy/core/tests/test_multiarray.py:40]: Honestly, not sure, as cool as pytest is, I guess its sometimes just too smart for me.

**Comment by eric-wieser** [comment]
Needs a rebase to clear out the first two commits

File: numpy/core/tests/test_conversion_utils.py [hunk 3]
```diff
@@ -1,45 +1,86 @@
...
+        sup = suppress_warnings()
+
+        if not self.exact_match:
+            if len(val) == 1:
+                # e.g. order
+                warn_msg = "case insensitive matches"
+            else:
+                # e.g. searchside, sortkind, clipmode
+                warn_msg = "inexact matches and case insensitive matches"
+            sup.filter(DeprecationWarning, warn_msg)
```
Comment (eric-wieser) [inline numpy/core/tests/test_conversion_utils.py:37]: Ideally you would use `pytest.warns` once for every relevant call to `self.conv`, to test _that_ it warns, not just to ignore when it warns.

Comment (anirudh2290) [reply numpy/core/tests/test_conversion_utils.py]: I replaced it with `assert_warns`, seems like probably due to pytest.ini settings, the DeprecationWarning still surface as errors, so pytest.warns doesnt catch it.

**Comment by anirudh2290** [comment]
1. I  deprecated byteorder inexact and case insensitive matches. For byteorder, I see that the test used 'big', 'little' etc. I changed it to 'B', 'L' etc. and deprecated the other non symbolic usages ('b' or 'big' or 'biggish'). 
2. Removed the doc in certain places which have deprecated usage and added a deprecation note.
3. Modified the tests and code which have deprecated usage.

File: numpy/core/src/multiarray/conversion_utils.c [hunk 4]
```diff
@@ -448,11 +474,24 @@ static int sortkind_parser(char const *str, Py_ssize_t length, void *data)
...
+
+    /* Filters out the case sensitive/non-exact
+     * match inputs and other inputs and outputs DeprecationWarning */
+    if (!((length == 4 && strcmp(str, "heap") == 0) ||
+          (length == 5 && strcmp(str, "quick") == 0) ||
+          (length == 6 && strcmp(str, "stable") == 0) ||
+          (length == 5 && strcmp(str, "merge") == 0)) && success == 0) {
+        /* NumPy 1.19, 2020-04-21 */
+        if (DEPRECATE("inexact matches and case insensitive matches "
+                      "for search side are deprecated. Please use "
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:488]:
```suggestion
                      "for sort kind are deprecated. Please use "
```

File: numpy/core/_add_newdocs.py [hunk 5]
```diff
@@ -3249,9 +3249,13 @@
...
         The default value ('S') results in swapping the current
-        byte order. The code does a case-insensitive check on the first
-        letter of `new_order` for the alternatives above.  For example,
-        any of 'B' or 'b' or 'biggish' are valid to specify big-endian.
+        byte order.
+
+       .. deprecated:: 1.19.0
+
+           Case insensitive and inexact matches are deprecated for new_order.
+           For example, any of 'big', 'biggish' or 'b' in place of 'B' are
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:3257]: `big` and `little` are consistent with `sys.byteorder`, so I don't think we want to deprecate them.

Comment (anirudh2290) [reply numpy/core/_add_newdocs.py:3257]: I did this since some of the numpy documentation uses 'B' 'L' etc. for byte order : https://numpy.org/doc/stable/reference/generated/numpy.ndarray.newbyteorder.html?highlight=numpy%20ndarray%20newbyteorder#numpy.ndarray.newbyteorder and I felt deprecating such usage may have more impact on our existing users than in other deprecations. What do you suggest here ? Should we skip byteorder deprecation or just deprecate non exact matches of 'big', 'little' etc.

**Comment by seberg** [comment]
Ah, I had missed the sortkind proposal. I am not sure I like it. First of all, we currently have `quicksort` in the documentation so I think that is good enough reason that we should keep it supported.
The question is whether we want to change the documentation to prefer the shorthand `quick`, but it seems to me that `"quick"` may be misleading, since it may be the slower sorter while "quicksort" is the name of the algorithm (for those who know it). (if it was `algoirthm=` and not `kind=` I would maybe think differently)

In other words, my first gut feeling would be to leave it as the current docs do it. I am happy to keep supporting `quick` and/or `q` as well though.

**Comment by anirudh2290** [comment]
> The question is whether we want to change the documentation to prefer the shorthand `quick`, but it seems to me that `"quick"` may be misleading, since it may be the slower sorter while "quicksort" is the name of the algorithm (for those who know it). (if it was `algoirthm=` and not `kind=` I would maybe think differently)

I agree 'quick' vs 'quicksort' can cause some confusion. I think I will revert my changes for sortkind and byteorder for now and just keep the deprecation limited to mode, order and searchside for now.

**Comment by seberg** [comment]
@anirudh2290 now that 1.19 is branched may be a good time to finish this. Just scrolled through, seems at least the "quick" documentation changes are still there.

File: numpy/core/_add_newdocs.py [hunk 6]
```diff
@@ -5652,15 +5649,11 @@
         byte order.  `new_order` codes can be any of:
 
         * 'S' - swap dtype from current to opposite endian
-        * {'<', 'L'} - little endian
-        * {'>', 'B'} - big endian
-        * {'=', 'N'} - native order
+        * {'<'} - little endian
+        * {'>'} - big endian
+        * {'='} - native order
         * {'|', 'I'} - ignore (no change to byte order)
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:5655]: Seems inconsistent to use sets to describe things with one element.

File: numpy/core/src/multiarray/conversion_utils.c [hunk 7]
```diff
@@ -373,8 +373,7 @@ static int byteorder_parser(char const *str, Py_ssize_t length, void *data)
 
     if (length < 1) {
         return -1;
-    }
-    else if (str[0] == NPY_BIG || str[0] == NPY_LITTLE
+    } else if (str[0] == NPY_BIG || str[0] == NPY_LITTLE
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:376]: I don't think we usually use this code style.

**Comment by anirudh2290** [comment]
> Just scrolled through, seems at least the "quick" documentation changes are still there.

Thanks @seberg for pointing this out, for some reason I thought this was done. will check again .

File: numpy/core/_add_newdocs.py (hunk 6 above)
Comment (anirudh2290) [reply numpy/core/_add_newdocs.py:5655]: sure good point, will change this

File: numpy/core/src/multiarray/conversion_utils.c (hunk 7 above)
Comment (anirudh2290) [reply numpy/core/src/multiarray/conversion_utils.c:376]: will do, thanks for pointing. separate discussion: would be nice to have a static analysis tool running style checks for these type of issues in the CI.

File: numpy/core/src/multiarray/conversion_utils.c [hunk 8]
```diff
@@ -497,21 +498,33 @@ PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind)
...
     NPY_SEARCHSIDE *side = (NPY_SEARCHSIDE *)data;
+    int success = -1;
 
     if (length < 1) {
         return -1;
     }
     else if (str[0] == 'l' || str[0] == 'L') {
         *side = NPY_SEARCHLEFT;
-        return 0;
+        success = 0;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:508]: I think this would be clearer as:
```
is_exact = (length == 4 && strcmp(str, "left") == 0);
```
then you can avoid doing multiple `strcmp`s when you only really need one.

You can keep the `-1` return in the error case, and then the if below becomes simply `if (!is_exact)`

File: numpy/core/src/multiarray/conversion_utils.c [hunk 9]
```diff
@@ -528,28 +541,43 @@ PyArray_SearchsideConverter(PyObject *obj, void *addr)
...
-    else {
+        success = 0;
+    } else {
         return -1;
     }
+
+    /* Filters out the case sensitive/non-exact
+     * match inputs and other inputs and outputs DeprecationWarning */
+    if (!((strcmp(str, "C") == 0 || strcmp(str, "F") == 0 ||
+           strcmp(str, "A") == 0 || strcmp(str, "K") == 0)) && success == 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:572]:
```suggestion
    if (!is_exact) {
```

File: numpy/core/src/multiarray/conversion_utils.c [hunk 10]
```diff
@@ -528,28 +541,43 @@ PyArray_SearchsideConverter(PyObject *obj, void *addr)
...
+    int success = -1;
+
     if (length != 1) {
         return -1;
     }
+
     if (str[0] == 'C' || str[0] == 'c') {
         *val = NPY_CORDER;
-        return 0;
+        success = 0;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:552]:
```suggestion
        is_exact = str[0] == 'c';
```

File: numpy/core/src/multiarray/conversion_utils.c [hunk 11]
```diff
@@ -497,21 +498,33 @@ PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind)
...
+
+    /* Filters out the case sensitive/non-exact
+     * match inputs and other inputs and outputs DeprecationWarning */
+    if (!((length == 4 && strcmp(str, "left") == 0) ||
+          (length == 5 && strcmp(str, "right") == 0)) && success == 0) {
+        /* NumPy 1.19, 2020-04-21 */
+        if (DEPRECATE("inexact matches and case insensitive matches "
+                      "for search side are deprecated. Please use "
+                      "one of 'left' or 'right' instead") < 0) {
+            success = -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:523]: You should check that `string_converter_helper` is actually able to handle exceptions being set. It might be replacing them, which it should not do.

File: numpy/core/tests/test_conversion_utils.py [hunk 12]
```diff
@@ -7,19 +7,31 @@
...
 
     def _check_value_error(self, val):
         pattern = r'\(got {}\)'.format(re.escape(repr(val)))
         with pytest.raises(ValueError, match=pattern) as exc:
             self.conv(val)
 
+    def _check_conv_assert_warn(self, val, expected):
+        if self.warn:
+            with assert_warns(DeprecationWarning) as exc:
+                assert self.conv(val) == expected
```
Comment (eric-wieser) [inline numpy/core/tests/test_conversion_utils.py:27]: Since this is a C api, it would be good to try making warnings errors, and catching that here too.

File: numpy/core/src/multiarray/conversion_utils.c (hunk 11 above)
Comment (anirudh2290) [reply numpy/core/src/multiarray/conversion_utils.c:523]: yes currently deprecation warnings surface as value errors. I will look into this.

File: numpy/core/tests/test_conversion_utils.py (hunk 12 above)
Comment (anirudh2290) [reply numpy/core/tests/test_conversion_utils.py:27]: I think this is covered by the tests in `numpy/core/tests/test_deprecations.py` but it surfaces as ValueError instead of DeprecationWarning. I will look into why string_converter_helper does this.

File: numpy/core/src/multiarray/conversion_utils.c [hunk 13]
```diff
@@ -570,24 +595,37 @@ PyArray_OrderConverter(PyObject *object, NPY_ORDER *val)
...
         *val = NPY_WRAP;
-        return 0;
+        is_exact = (length == 4 && strcmp(str, "wrap") == 0);
     }
     else if (str[0] == 'R' || str[0] == 'r') {
         *val = NPY_RAISE;
-        return 0;
+        is_exact = (length == 5 && strcmp(str, "raise") == 0);
     }
-    else {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:588]: I don't think you want to remove this

File: numpy/core/src/multiarray/conversion_utils.c [hunk 14]
```diff
@@ -528,28 +539,42 @@ PyArray_SearchsideConverter(PyObject *obj, void *addr)
...
-        return 0;
+        is_exact = str[0] == 'A';
     }
     else if (str[0] == 'K' || str[0] == 'k') {
         *val = NPY_KEEPORDER;
-        return 0;
-    }
-    else {
+        is_exact = str[0] == 'K';
+    } else {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:563]: style: let's leave this line as it was,
```suggestion
    }
    else {
```

File: numpy/core/src/multiarray/conversion_utils.c [hunk 15]
```diff
@@ -497,21 +497,32 @@ PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind)
...
-        return 0;
+        is_exact = (length == 4 && strcmp(str, "left") == 0);
     }
     else if (str[0] == 'r' || str[0] == 'R') {
         *side = NPY_SEARCHRIGHT;
-        return 0;
+        is_exact = (length == 5 && strcmp(str, "right") == 0);
     }
-    else {
-        return -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:513]: This line needs to come back

File: numpy/core/src/multiarray/conversion_utils.c (hunk 13 above)
Comment (anirudh2290) [reply numpy/core/src/multiarray/conversion_utils.c:588]: aah! i think i should probably add a test for this use case.

File: numpy/core/src/multiarray/conversion_utils.c (hunk 11 above)
Comment (anirudh2290) [reply numpy/core/src/multiarray/conversion_utils.c]: I have addressed this issue, I skip setting ValueError if error already occured.

File: numpy/core/tests/test_conversion_utils.py (hunk 12 above)
Comment (anirudh2290) [reply numpy/core/tests/test_conversion_utils.py:27]: since `numpy/core/tests/test_deprecations.py` already covers raising warnings as errors, can we keep test_conversion_utils as it is ?

File: numpy/core/src/multiarray/conversion_utils.c [hunk 16]
```diff
@@ -497,21 +499,34 @@ PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind)
...
-        return 0;
+        is_exact = (length == 4 && strcmp(str, "left") == 0);
     }
     else if (str[0] == 'r' || str[0] == 'R') {
         *side = NPY_SEARCHRIGHT;
-        return 0;
-    }
-    else {
+        is_exact = (length == 5 && strcmp(str, "right") == 0);
+    } else {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:514]:
```suggestion
    }
    else {
```

Comment (anirudh2290) [reply numpy/core/src/multiarray/conversion_utils.c]: thanks, i have undone these changes.

**Comment by anirudh2290** [comment]
I have addressed @eric-wieser 's comments above and added a release note. This is ready for review.

File: numpy/core/_add_newdocs.py [hunk 17]
```diff
@@ -3243,15 +3243,13 @@
         below. `new_order` codes can be any of:
 
         * 'S' - swap dtype from current to opposite endian
-        * {'<', 'L'} - little endian
-        * {'>', 'B'} - big endian
-        * {'=', 'N'} - native order
+        * '<' - little endian
+        * '>' - big endian
+        * '=' - native order
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:3248]: Do we want to explicitly indicate we accept the `sys.byteorder` names?
```suggestion
        * {'<', 'little'} - little endian
        * {'>', 'big'} - big endian
        * '=' - native order, equivalent to `sys.byteorder`
```

Comment (anirudh2290) [reply numpy/core/_add_newdocs.py:3248]: not sure, sys.byteorder means only 'big' or 'little' and this list is a superset of that. I dont think it has any negative effect though.

File: doc/release/upcoming_changes/16056.deprecation.rst [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+-------------------------------------------------------------
+Inexact and case insensitive matches for mode, order and searchside were
+valid inputs earlier and will give a DeprecationWarning now.
+For example, below are some example usages which are now deprecated and will
+give a DeprecationWarning.
+
+    import numpy as np
+    arr = np.array([[3, 6, 6], [4, 5, 1]])
+    # order: case insensitive match
+    np.ravel_multi_index(arr, (7, 6), order="false")
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/16056.deprecation.rst:11]: This was never legal, was it?
```suggestion
    np.ravel_multi_index(arr, (7, 6), order="f")
```

File: doc/release/upcoming_changes/16056.deprecation.rst [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+    # mode: inexact match
+    np.ravel_multi_index(arr, (7, 6), mode="clap")
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/16056.deprecation.rst:13]:
```suggestion
    np.ravel_multi_index(arr, (7, 6), mode="clap")  # should be "clip"
```

File: doc/release/upcoming_changes/16056.deprecation.rst [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+    # searchside: inexact match
+    np.searchsorted(arr[0], 4, side='random')
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/16056.deprecation.rst:15]:
```suggestion
    np.searchsorted(arr[0], 4, side='random')  # should be "right"
```

File: numpy/core/_add_newdocs.py [hunk 21]
```diff
@@ -5652,15 +5650,11 @@
         byte order.  `new_order` codes can be any of:
 
         * 'S' - swap dtype from current to opposite endian
-        * {'<', 'L'} - little endian
-        * {'>', 'B'} - big endian
-        * {'=', 'N'} - native order
+        * '<' - little endian
+        * '>' - big endian
+        * '=' - native order
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:5655]: Can we make this match the one above after my suggestions?

File: numpy/core/_add_newdocs.py [hunk 22]
```diff
@@ -6260,19 +6254,17 @@
     The `new_order` code can be any from the following:
 
     * 'S' - swap dtype from current to opposite endian
-    * {'<', 'L'} - little endian
-    * {'>', 'B'} - big endian
-    * {'=', 'N'} - native order
+    * '<' - little endian
+    * '>' - big endian
+    * '=' - native order
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:6259]: And this one too

File: numpy/core/src/multiarray/conversion_utils.c [hunk 23]
```diff
@@ -360,8 +360,10 @@ string_converter_helper(
     int ret = str_func(str, length, out);
     Py_DECREF(str_object);
     if (ret < 0) {
+        if (!PyErr_Occurred()) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:363]:
```suggestion
        /* str_func returns -1 without an exception if the value is wrong */
        if (!PyErr_Occurred()) {
```

File: numpy/core/src/multiarray/conversion_utils.c [hunk 24]
```diff
@@ -497,21 +499,35 @@ PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind)
...
         return -1;
     }
+
+    /* Filters out the case sensitive/non-exact
+     * match inputs and other inputs and outputs DeprecationWarning */
+    if (!is_exact) {
+        /* NumPy 1.20, 2020-05-19 */
+        if (DEPRECATE("inexact matches and case insensitive matches "
+                      "for search side are deprecated. Please use "
+                      "one of 'left' or 'right' instead") < 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/conversion_utils.c:525]:
```suggestion
                      "one of 'left' or 'right' instead.") < 0) {
```

Comment (eric-wieser) [reply numpy/core/src/multiarray/conversion_utils.c:525]: And below. Either than, or rephrase this to use `deprecate, please` instead of `deprecated. Please`.

File: numpy/core/tests/test_conversion_utils.py (hunk 12 above)
Comment (eric-wieser) [reply numpy/core/tests/test_conversion_utils.py:27]: Sure, that's reasonable.

File: doc/release/upcoming_changes/16056.deprecation.rst (hunk 18 above)
Comment (anirudh2290) [reply doc/release/upcoming_changes/16056.deprecation.rst:11]: aah good point, thanks !

File: doc/release/upcoming_changes/16056.deprecation.rst [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+-------------------------------------------------------------
+Inexact and case insensitive matches for mode, order and searchside were
+valid inputs earlier and will give a DeprecationWarning now.
+For example, below are some example usages which are now deprecated and will
+give a DeprecationWarning.
+
+    import numpy as np
+    arr = np.array([[3, 6, 6], [4, 5, 1]])
+    # order: case insensitive match
+    np.ravel_multi_index(arr, (7, 6), order="f")  # should be "F"
```
Comment (seberg) [inline doc/release/upcoming_changes/16056.deprecation.rst:11]: Nit: Maybe use `np.array()`, or `copy` which also have the order parameter and are more prominent.

File: numpy/core/src/multiarray/conversion_utils.c [hunk 26]
```diff
@@ -582,24 +613,39 @@ PyArray_OrderConverter(PyObject *object, NPY_ORDER *val)
...
         *val = NPY_RAISE;
-        return 0;
+        is_exact = (length == 5 && strcmp(str, "raise") == 0);
     }
     else {
         return -1;
     }
+
+    /* Filters out the case sensitive/non-exact
+     * match inputs and other inputs and outputs DeprecationWarning */
```
Comment (seberg) [inline numpy/core/src/multiarray/conversion_utils.c:638]: Nit, we usually use:
```
/*
 * multie line
 * comment
 */
```
at the moment

**Review by seberg — APPROVED** [review]
Thanks Anirudh, looks very nice. The only thing that makes me hesitate a bit is that I just checked scipy, and there are a few files that seem to use `order='c'` (it seems mainly two files, one file has `order="f"`), which might indicate that lower cases are used occasionally.

We may get away with it though, but could maybe fixup scipy before in either case? But if anyone has doubts, please say so, because otherwise one of these days I may decide to merge it :).

**Comment by mattip** [comment]
The [mailing list discussion](http://numpy-discussion.10968.n7.nabble.com/Deprecate-inexact-matches-for-mode-order-etc-td48100.html) did not generate much feedback, but I am concerned that this may bite downstream packages and users. Are we really sure that pandas/astropy/Dask have no problem with this?

**Comment by seberg** [comment]
Well, I checked those large packages for the ones that would be likely `order="f"` (with lower case) and `side="l"` as a shorthand, and only found 3 files in SciPy.

I would be happy with keeping the lower case, if we expect those 3 usage/files in SciPy indicate that it is common enough that we do not want to create churn.  That I feel would be the main point, unnecessary churn opposed to the advantage of unified style.

**Comment by mattip** [comment]
> I would be happy with keeping the lower case, if we expect those 3 usage/files in SciPy indicate that it is common enough that we do not want to create churn.

That would be good

**Comment by anirudh2290** [comment]
Thanks @seberg and @mattip . I have addressed the review comments and removed order from deprecation to avoid churn. Thanks for checking the downstream projects @seberg ! I also ran the tests for pandas, dask, scipy and didnt find anything except the usage that you mentioned.

File: doc/release/upcoming_changes/16056.deprecation.rst [hunk 27]
```diff
@@ -0,0 +1,13 @@
+Inexact matches for mode, order and searchside are deprecated
+-------------------------------------------------------------
+Inexact and case insensitive matches for mode, order and searchside were
```
Comment (seberg) [inline doc/release/upcoming_changes/16056.deprecation.rst:3]:
```suggestion
Inexact matches for mode and searchside are deprecated
------------------------------------------------------
Inexact and case insensitive matches for mode and searchside were
```

Thanks for persisting @anirudh2290. I will do a last pass today and then put this in I think!

Comment (anirudh2290) [reply doc/release/upcoming_changes/16056.deprecation.rst]: Thanks Sebastian!  Good catch, I have addressed your comments.

**Review by mattip — APPROVED** [review]

**Comment by seberg** [comment]
Thanks Anirudh! Let me put this in.
